### ISSUE TITLE
Bug: Template variables from chained conditions not populated in rule messages

### ISSUE THREAD

**Issue by tsanders-rh** [konveyor/analyzer-lsp#1000]
# Bug: Template variables from chained conditions not populated in rule messages

## Summary

When using rule chaining with `as:` to capture context from a condition, template variables like `{{asVar.Filepaths}}` are not populated in rule messages, despite the ChainTemplate being correctly created internally. This makes it impossible to reference chained condition results in violation messages.

## Environment

- **analyzer-lsp**: Latest version (tested 2024-11)
- **Provider**: nodejs.referenced (grpc provider)
- **Affects**: All providers using `as:` with template variables in messages

### Test Rule
```yaml
- ruleID: test-template-variable-bug
  description: Demonstrates template variable not being populated
  when:
    as: componentFiles
    nodejs.referenced:
      pattern: ToolbarGroup
  message: "Found ToolbarGroup in files: {{componentFiles.Filepaths}}"
  effort: 1
```

### Expected Behavior
```
Found ToolbarGroup in files: /opt/input/source/app/file1.tsx,/opt/input/source/app/file2.tsx
```

### Actual Behavior
```
Found ToolbarGroup in files:
```

The template variable `{{componentFiles.Filepaths}}` is empty despite 17 incidents being found.

### 1. Incidents Are Returned With FileURI
The nodejs.referenced provider correctly returns incidents with FileURI set:

```yaml
incidents:
- uri: file:///opt/input/source/applications/analysis-wizard/custom-rules.tsx
  message: 'Found ToolbarGroup in files: '
  lineNumber: 18
  variables:
    file: file:///opt/input/source/applications/analysis-wizard/custom-rules.tsx
```

### 2. ChainTemplate Is Populated Correctly
At `engine/conditions.go:181-184`, the ChainTemplate is created correctly:

```go
if c.As != "" {
    condCtx.Template[c.As] = ChainTemplate{
        Filepaths: incidentsToFilepaths(response.Incidents),  // ← Correctly extracts paths
        Extras:    response.TemplateContext,
    }
}
```

### 3. incidentsToFilepaths() Works Correctly
At `engine/conditions.go:278-288`, file paths are extracted from incidents:

```go
func incidentsToFilepaths(incident []IncidentContext) []string {
    filepaths := []string{}
    for _, ic := range incident {
        if strings.HasPrefix(string(ic.FileURI), uri.FileScheme) {
            filepaths = append(filepaths, ic.FileURI.Filename())
        } else {
            filepaths = append(filepaths, string(ic.FileURI))
        }
    }
    return filepaths
}
```

## Root Cause

**Location**: `engine/engine.go:602-749` in `createViolation()`, specifically lines 695-708

**Problem**: The `ConditionContext.Template` data is never passed to message rendering.

### Issue #1: createViolation Doesn't Receive ConditionContext

At `engine/engine.go:301`:
```go
violation, err := r.createViolation(ctx, response.ConditionResponse, response.Rule, scopes)
```

The function signature (line 602):
```go
func (r *ruleEngine) createViolation(ctx context.Context, conditionResponse ConditionResponse, rule Rule, scope Scope) (konveyor.Violation, error)
```

**Missing**: The `ConditionContext` parameter that contains the `Template` map.

### Issue #2: Message Rendering Only Uses Incident Variables

At `engine/engine.go:695-708`:
```go
if rule.Perform.Message.Text != nil {
    variables := make(map[string]interface{})
    for key, value := range m.Variables {
        variables[key] = value  // ← Only copies incident-specific variables
    }
    if m.LineNumber != nil {
        variables["lineNumber"] = *m.LineNumber
    }
    // ❌ Missing: No code to add condCtx.Template to variables map
    templateString, err := r.createPerformString(*rule.Perform.Message.Text, variables)
    if err != nil {
        r.logger.Error(err, "unable to create template string")
    }
    incident.Message = templateString
}
```

The `variables` map passed to `createPerformString()` (which calls `mustache.Render()`) only contains:
- Incident-specific variables from `m.Variables`
- `lineNumber` if present

**It does NOT contain the chain template data from `condCtx.Template`.**

## Data Flow Diagram

```
1. Condition evaluates (e.g., nodejs.referenced)
   ↓
2. condCtx.Template["componentFiles"] = ChainTemplate{
     Filepaths: [file1.tsx, file2.tsx, ...],  ✅ Populated correctly
     Extras: {...}
   }
   ↓
3. ConditionResponse returned
   - Contains: Incidents ✅
   - Missing: Template data ❌
   ↓
4. createViolation() called at engine.go:301
   - Receives: ConditionResponse ✅
   - Missing: ConditionContext with Template ❌
   ↓
5. Message rendering at engine.go:703
   - Uses: m.Variables (incident vars only) ✅
   - Missing: condCtx.Template (chain context) ❌
   ↓
6. mustache.Render() cannot resolve {{componentFiles.Filepaths}}
   ↓
7. Result: Empty template variable in message
```

### Option 1: Pass ConditionContext to createViolation (Recommended)

**Modify** `engine/engine.go:301` and similar calls:
```go
// Current (lines 288-301):
case response := <-ret:
    func() {
        // ... code ...
        violation, err := r.createViolation(ctx, response.ConditionResponse, response.Rule, scopes)
        // ...
    }()

// Proposed:
case response := <-ret:
    func() {
        // ... code ...
        violation, err := r.createViolation(ctx, response.ConditionResponse, response.Rule, scopes, response.ConditionContext)
        // ...
    }()
```

**Update function signature** at `engine/engine.go:602`:
```go
func (r *ruleEngine) createViolation(
    ctx context.Context,
    conditionResponse ConditionResponse,
    rule Rule,
    scope Scope,
    conditionContext ConditionContext,  // ← Add this parameter
) (konveyor.Violation, error)
```

**Merge Template data** into variables at `engine/engine.go:695-708`:
```go
if rule.Perform.Message.Text != nil {
    variables := make(map[string]interface{})

    // Add chain template data
    for key, chainTemplate := range conditionContext.Template {
        variables[key] = map[string]interface{}{
            "Filepaths": chainTemplate.Filepaths,
            "Extras":    chainTemplate.Extras,
        }
    }

    // Add incident-specific variables
    for key, value := range m.Variables {
        variables[key] = value
    }
    if m.LineNumber != nil {
        variables["lineNumber"] = *m.LineNumber
    }

    templateString, err := r.createPerformString(*rule.Perform.Message.Text, variables)
    // ...
}
```

### Option 2: Include Template in Response

Modify the `response` struct (line 58-63) to include ConditionContext:
```go
type response struct {
    ConditionResponse ConditionResponse `yaml:"conditionResponse"`
    Err               error             `yaml:"err"`
    Rule              Rule              `yaml:"rule"`
    RuleSetName       string
    ConditionContext  ConditionContext  `yaml:"conditionContext"` // ← Add this
}
```

Then populate it where responses are created (e.g., at `engine/engine.go:188-195`).

## Impact

This bug affects:
- ✅ **All** chained conditions using `as:` with template variables in messages
- ✅ **All** grpc providers (nodejs.referenced, java.referenced, etc.)
- ✅ Rule chaining for filtering subsequent conditions by files
- ✅ Displaying context from previous conditions in violation messages

## Workaround

Currently, users must use 3-step `builtin.filecontent` chains instead of `nodejs.referenced`:

```yaml
# Step 1: Find component usage (no message, only for chaining)
- ruleID: find-toolbar-group-usage
  when:
    as: componentFiles
    builtin.filecontent:
      pattern: "import.*ToolbarGroup.*@patternfly/react-core"
  message: ""  # Workaround: Can't use {{componentFiles.Filepaths}} here

# Step 2: Verify imports in those files
- ruleID: verify-toolbar-group-imports
  when:
    from: componentFiles.Filepaths
    as: imports
    builtin.filecontent:
      pattern: "import.*ToolbarGroup.*@patternfly/react-core"

# Step 3: Final check with message
- ruleID: toolbar-group-violation
  when:
    from: imports.Filepaths
    builtin.filecontent:
      pattern: "<ToolbarGroup"
  message: "ToolbarGroup component usage found - migrate to ToolbarItem"
```

This workaround works because `builtin.filecontent` correctly populates template variables, but it's verbose and inefficient compared to direct `nodejs.referenced` chaining.

## Test Case

See attached test file that demonstrates the bug:

<details>
<summary>test-template-debug.yaml</summary>

```yaml
- name: konveyor-analysis
  description: Test case for template variable bug
  labels:
  - konveyor.io/source=patternfly-v5
  - konveyor.io/target=patternfly-v6
  rules:
  - ruleID: test-template-debug-00000
    description: Debug test - nodejs.referenced with 'as' to see what gets populated
    effort: 1
    category: potential
    labels:
    - konveyor.io/source=patternfly-v5
    - konveyor.io/target=patternfly-v6
    when:
      as: componentFiles
      nodejs.referenced:
        pattern: ToolbarGroup
    message: "Found ToolbarGroup - template should contain: {{componentFiles.Filepaths}}"
```
</details>

## Files Affected

- `engine/engine.go` (lines 58-63, 188-195, 301, 466, 602-749)
- `engine/conditions.go` (lines 175-201, 278-288)

## Additional Context

This issue was discovered while developing migration rules for PatternFly v5 to v6. The inability to use template variables from chained conditions significantly limits the expressiveness of rules and forces workarounds that are less efficient and harder to maintain.

Related functionality that DOES work:
- ✅ Template variables from same condition (e.g., `{{file}}`, `{{lineNumber}}`)
- ✅ Template variables in tag strings (tags use incident.Variables correctly)
- ✅ Chaining with `from:` to filter subsequent conditions by filepath
- ✅ `builtin.filecontent` provider template variable population

Thank you for maintaining this project! Happy to provide additional information or test any proposed fixes.
